File: typescript/longest-palindromic-substring.ts
function longestPalindrome(s: string): string {
  let longestPalindromeInterval: [number, number] = [0, 0];
  for (let i = 0; i < s.length; i++) {
    let [from, to] = longestPalindromeFromIndex(s, i);
    if (
      to - from > longestPalindromeInterval[1] - longestPalindromeInterval[0]
    ) {
      longestPalindromeInterval = [from, to];
    }
    [from, to] = longestPalindromeFromIndex(s, i, i + 1);
    if (
      to - from > longestPalindromeInterval[1] - longestPalindromeInterval[0]
    ) {
      longestPalindromeInterval = [from, to];
    }
  }
  return s.slice(...longestPalindromeInterval);
}

function longestPalindromeFromIndex(
  str: string,
  index1: number,
  index2: number = index1,
): [number, number] {
  let i = index1;
  let j = index2;
  while (i >= 0 && j < str.length && str[i] === str[j]) {
    i -= 1;
    j += 1;
  }

  return [i + 1, j];
}

import { assertArrayIncludes, assertEquals } from "std/assert/mod.ts";

Deno.test("examples", () => {
  assertEquals(longestPalindrome("cbbd"), "bb");
  assertArrayIncludes(["bab", "aba"], [longestPalindrome("babad")]);
  assertArrayIncludes(["a", "b", "c", "d"], [longestPalindrome("abcd")]);
});

File: python3/longest-palindromic-substring.py
class Solution:
    """
    >>> Solution().longestPalindrome("babad") in ("bab", "aba")
    True
    >>> Solution().longestPalindrome("cbbd")
    'bb'
    """

    def longestPalindrome(self, s: str) -> str:
        def find_palindrome(i, j) -> tuple[int, int]:
            while 0 <= i and j < len(s) and s[i] == s[j]:
                i -= 1
                j += 1
            i += 1
            j -= 1
            return (i, j)
        longest = 0
        longest_start, longest_end = None, None
        for i, _ in enumerate(s):
            for j in (i, i + 1):
                start, end = find_palindrome(i, j)
                length = end - start + 1
                if length > longest:
                    longest = length
                    longest_start, longest_end = start, end
        return s[longest_start:longest_end + 1]
